src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1242,19 +1246,192 @@
+                        prev)
+            :test (handle-expr prev
+                               (compile-test (:constraints condition))
+                               id
+                               :test-expr
+                               {:compile-ctx {:condition condition
+                                              :env (:env beta-node)
+                                              :msg "compiling test node"}})
+            :accumulator (cond-> (handle-expr prev
+                                              (compile-accum (:accumulator beta-node) (:env beta-node))
+                                              id
+                                              :accum-expr
+                                              {:compile-ctx {:condition condition
+                                                             :accumulator (:accumulator beta-node)
+                                                             :env (:env beta-node)
+                                                             :msg "compiling accumulator"}})
+
+                                 (:join-filter-expressions beta-node)
+                                 (handle-expr (compile-join-filter (:join-filter-expressions beta-node)
+                                                                   (:join-filter-join-bindings beta-node)
+                                                                   (:new-bindings beta-node)
+                                                                   (:env beta-node))
+                                              id
+                                              :join-filter-expr
+                                              {:compile-ctx {:condition condition
+                                                             :join-filter-expressions (:join-filter-expressions beta-node)
+                                                             :env (:env beta-node)
+                                                             :msg "compiling accumulate with join filter node"}}))
+            :query prev)))
+      exprs
+      (:id-to-condition-node beta-graph))))
+
+(sc/defn compile-exprs :- schema/NodeFnLookup
+  "Takes a map in the form produced by extract-exprs and evaluates the values(expressions) of the map in a batched manner.
+   This allows the eval calls to be more effecient, rather than evaluating each expression on its own.
+   See #381 for more details."
+  [key->expr :- schema/NodeExprLookup
+   partition-size :- sc/Int]
+  (let [batching-try-eval (fn [node-keys exprs]
+                            ;; Try to evaluate all of the expressions as a batch. If they fail the batch eval then we
+                            ;; try them one by one with their compilation context, this is slow but we were going to fail
+                            ;; anyway.
+                            (try
+                              (eval exprs)
+                              (catch Exception e
```

Comment:
I wonder if there are scenarios where the compilation could throw an Error other than a VirtualMachineError (which it wouldn't make sense to try to recover from, see the [docs](https://docs.oracle.com/javase/7/docs/api/java/lang/VirtualMachineError.html))

Reply:
`catch Exception` shouldn't catch these sorts of errors anyway right? 

But you do make a good point, [InterruptedException](https://docs.oracle.com/javase/7/docs/api/java/lang/InterruptedException.html) would be a good candidate to not catch

Reply:
TODO: determine what exceptions to not catch

Reply:
I don't see a reason why you shouldn't catch all exceptions and rethrow here. I think that the message needs to be clear that it could only "possibly" be a problem with compilation of the expression. However, I don't think it is beneficial to rethrow other exceptions in their "raw" form . It is more useful to wrap the exception and provide it with what context you can. Attach the root cause exception though so it is bubbles up.

More context should always be a better thing in this sort of case.

Reply:
I was actually musing on whether there are any Error types that we should catch, with the idea being to rethrow them with the specific expression that caused them rather than a gigantic collections that probably isn't too useful to the user.  I don't know that the traditional assumptions about not recovering from Error necessarily apply to a dynamic code generation and compilation framework - I particularly had [LinkageError](https://docs.oracle.com/javase/7/docs/api/java/lang/LinkageError.html) in mind.  We definitely wouldn't want to catch any VirtualMachineError instances though since the JVM's behavior when you do is specifically called out as undefined by the spec.

In practice this probably isn't a huge deal and I don't want to block this on investigating it - I'd expect most bad cases to throw Exception.  Maybe just a followup issue to think about later.